### PR TITLE
Feature/nodelines

### PR DESCRIPTION
**Pull request by liuloppan**
Function that allows creating lines between nodes. 
The purpose is to use this to illustrate the distances between planets (with some complementary label), but in the future it may be used as a base class for other types of line rendering or simply used as a debugging tool.

### CODE REVIEW

File: modules/base/rendering/renderablenodeline.cpp [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+#include <modules/base/basemodule.h>
+#include <openspace/documentation/verifier.h>
+#include <openspace/engine/globals.h>
+#include <openspace/rendering/renderengine.h>
+#include <openspace/scene/scene.h>
+#include <openspace/scene/translation.h>
+#include <openspace/util/updatestructures.h>
+#include <ghoul/filesystem/filesystem.h>
+#include <ghoul/opengl/programobject.h>
+#include <openspace/interaction/navigationhandler.h>
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: Include ordering;  navigationhandler and orbitalnavigator should be grouped (and alphabetically sorted) with the other OpenSpace includes

File: modules/base/rendering/renderablenodeline.cpp [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+#include <openspace/interaction/orbitalnavigator.h>
+
+namespace {
+
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: No need for the extra empty line here

File: modules/base/rendering/renderablenodeline.cpp [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+        "Color",
+        "This value determines the RGB color for the line."
+    };
+
+    constexpr openspace::properties::Property::PropertyInfo LineWidthInfo = {
+        "LineWidth",
+        "Line Width",
+        "This value specifies the line width."
+    };
+
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: No need for the extra empty line here

File: modules/base/rendering/renderablenodeline.cpp [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+    );
+
+    _start = dictionary.value<std::string>(StartNodeInfo.identifier);
+    _end = dictionary.value<std::string>(EndNodeInfo.identifier);
+
+    if (dictionary.hasKey(LineColorInfo.identifier)) {
+        _lineColor = dictionary.value<glm::vec3>(LineColorInfo.identifier);
+    }
+    if (dictionary.hasKey(LineWidthInfo.identifier)) {
+        _lineWidth = float(dictionary.value<double>(LineWidthInfo.identifier));
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: Prefer C++ style casts over C casts:
`static_cast<float>(dictionary.value<double>(LineWidthInfo.identifier));`

Reasoning:  A C-cast can mean many different things and it requires some thinking to figure out what is meant in each situation.  The C++ style makes it more clear and prevent bugs from being introduced in general

File: modules/base/rendering/renderablenodeline.cpp [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        }
+    );
+
+    // Generate
+    glGenVertexArrays(1, &_vaoId);
+    glGenBuffers(1, &_vBufferId);
+
+    bindGL();
+
+    glVertexAttribPointer(_locVertex, 3, GL_FLOAT, GL_FALSE, 3 * sizeof(float), (void*)0);
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: can use `nullptr` as the last parameter which looks a bit nicer

File: modules/base/rendering/renderablenodeline.cpp [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+    glEnableVertexAttribArray(_locVertex);
+
+    unbindGL();
+}
+
+void RenderableNodeLine::deinitializeGL() {
+
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: Extra empty line

File: modules/base/rendering/renderablenodeline.cpp [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+    glDeleteVertexArrays(1, &_vaoId);
+    _vaoId = 0;
+
+    glDeleteBuffers(1, &_vBufferId);
+    _vBufferId = 0;
+
+    BaseModule::ProgramObjectManager.release(
+        ProgramName,
+        [](ghoul::opengl::ProgramObject* p) {
+        global::renderEngine.removeRenderProgram(p);
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: This and the next line are missing an indentation level (thanks Visual Studio for autocorrecting this *all the time*)

File: modules/base/rendering/renderablenodeline.cpp [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+    }
+    );
+    _program = nullptr;
+}
+
+bool RenderableNodeLine::isReady() const {
+    bool ready = true;
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: Can replace with `return _program != nullptr`

File: modules/base/rendering/renderablenodeline.cpp [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+    ready &= (_program != nullptr);
+    return ready;
+}
+
+void RenderableNodeLine::unbindGL() {
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: Might be better to remove the `unbindGL` and `bindGL` functions and make the calls for binding directly.

Also:  No need for the `glBindBuffer` calls.  If you bind or unbind a vertex array object, it's associated vertex buffer is unbound as well.

File: modules/base/rendering/renderablenodeline.cpp [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+    glBindBuffer(GL_ARRAY_BUFFER, 0);
+    glBindVertexArray(0);
+}
+
+void RenderableNodeLine::bindGL()
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: Prefer `{` on the same line as the function

File: modules/base/rendering/renderablenodeline.cpp [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+{
+    glBindVertexArray(_vaoId);
+    glBindBuffer(GL_ARRAY_BUFFER, _vBufferId);
+}
+
+void RenderableNodeLine::updateVertexData()
+{
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: Same as above

File: modules/base/rendering/renderablenodeline.cpp [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+
+    _vertexArray.push_back(_startPos.x);
+    _vertexArray.push_back(_startPos.y);
+    _vertexArray.push_back(_startPos.z);
+
+    _vertexArray.push_back(_endPos.x);
+    _vertexArray.push_back(_endPos.y);
+    _vertexArray.push_back(_endPos.z);
+
+    _vertexArray;
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: This line doesn't do anything

File: modules/base/rendering/renderablenodeline.cpp [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+    );
+
+    //update vertex attributes
+    glVertexAttribPointer(_locVertex, 3, GL_FLOAT, GL_FALSE, 3 * sizeof(float), nullptr);
+
+    unbindGL();
+}
+
+void RenderableNodeLine::render(const RenderData& data, RendererTasks&) {
+
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: Extra empty line

File: modules/base/rendering/renderablenodeline.cpp [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+    if (!isBlendEnabled) {
+        glDisablei(GL_BLEND, 0);
+    }
+    if (!isLineSmoothEnabled) {
+        glDisable(GL_LINE_SMOOTH);
+    }
+}
+
+
+/*  Returns a position that is relative to the current
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: prefer `//` style of comments

Reasoning:  If only `//` stlye of comments are used in a file, it's possible to comment out large parts of it by using `/*  */` temporarily.  If `/*` is used instead, that doesn't work as:
```
/*
  commented out
  /*
    still commented out
  */
   *not* commented out
*/   Compile error
```

File: modules/base/rendering/renderablenodeline.cpp [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+    anchor node. This is a method to handle precision
+    problems that occur when approaching a line end point */
+glm::dvec3 RenderableNodeLine::getCoordinatePosFromAnchorNode(glm::dvec3 worldPos) {
+
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: Extra empty line

File: modules/base/rendering/renderablenodeline.cpp [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+    glm::dvec3 anchorNodePos(0);
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: Use `0.0` to initialize double variable

File: modules/base/rendering/renderablenodeline.cpp [hunk 17]
```diff
@@ -0,0 +1,307 @@
...
+
+    if (global::navigationHandler.orbitalNavigator().anchorNode()) {
+        anchorNodePos = global::navigationHandler.orbitalNavigator().anchorNode()->worldPosition();
+    }
+    glm::dvec3 diffPos = glm::dvec3(worldPos.x - anchorNodePos.x, worldPos.y - anchorNodePos.y,
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: Can be replaced with:
`glm::dvec3 diffPos = worldPos - anchorNodePos);`

File: modules/base/rendering/renderablenodeline.cpp [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+        worldPos.z - anchorNodePos.z);
+
+    return diffPos;
+}
+
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.cpp]: Nitpicky

Extra empty line here

File: modules/base/rendering/renderablenodeline.h [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+
+#include <openspace/rendering/renderable.h>
+
+#include <openspace/properties/scalar/boolproperty.h>
+#include <openspace/properties/scalar/floatproperty.h>
+#include <openspace/properties/vector/vec3property.h>
+#include <ghoul/opengl/ghoul_gl.h>
+#include <ghoul/glm.h>
+
+namespace ghoul::opengl {
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.h]: Prefer
`namespace ghoul::opengl { class ProgramObject; }`
for single objects

File: modules/base/rendering/renderablenodeline.h [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+ */
+class RenderableNodeLine : public Renderable {
+public:
+    RenderableNodeLine(const ghoul::Dictionary& dictionary);
+    ~RenderableNodeLine() = default;
+
+    static documentation::Documentation Documentation();
+    
+private:
+
```
Comment (alexanderbock) [inline modules/base/rendering/renderablenodeline.h]: Extra empty line

File: modules/base/shaders/line_vs.glsl [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+layout(location = 0) in vec3 in_position;
+
+out vec4 vs_positionScreenSpace;
+out vec4 vs_positionViewSpace;
+
+uniform mat4 modelViewTransform;
+uniform mat4 projectionTransform;
+
+void main() {
+
```
Comment (alexanderbock) [inline modules/base/shaders/line_vs.glsl]: Extra empty line

File: modules/base/shaders/line_vs.glsl [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+    vs_positionViewSpace = vec4(modelViewTransform * dvec4(in_position, 1));
+    vs_positionScreenSpace = projectionTransform * vs_positionViewSpace;
+    gl_Position  = vs_positionScreenSpace;
+
+    // Set z to 0 to disable near and far plane, unique handling for perspective in space
+    gl_Position.z = 0.f; 
+
```
Comment (alexanderbock) [inline modules/base/shaders/line_vs.glsl]: Extra empty line

File: modules/base/shaders/line_fs.glsl [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT    *
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF  *
+ * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE  *
+ * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                                         *
+ ****************************************************************************************/
+
+#include "fragment.glsl"
+#include "floatoperations.glsl"
+
+in vec4 vs_positionScreenSpace;
```
Comment (alexanderbock) [inline modules/base/shaders/line_fs.glsl]: If you only need the `w` component, passing only a `float` would suffice here

**Comment by liuloppan** [comment]
I applied the suggested changes.
Tested and works with ' SGCTConfig = sgct.config.fisheye{1024, 1024}'

**Pull request merged**